File: 4/solution.py
def const(a, b):
    def pair(f):
        return f(a, b)
    return pair

def car(pair):
    def first(a, b):
        return a
    return pair(first)

def cdr(pair):
    def last(a, b):
        return b
    return pair(last)

assert(car(const(5, 9)) == 5)
assert(cdr(const(3, 20)) == 20)

assert(car(const("a", "b")) == "a")
assert(cdr(const(3.5, 4.5)) == 4.5)

assert(car(const(3, 4)) == 3)
assert(cdr(const(3, 4)) == 4)

File: 4/solution.php
<?php

function lowest($input)
{
    sort($input);

    $last = null;

    foreach ($input as $number) {

        if ($number <= 0)
        {
            continue;
        }

        if ($last + 1 !== $number)
        {
            return $last + 1;
        }

        $last = $number;
    }

    return $last + 1;
}

assert(lowest([3, 4, -1, 1]) === 2);
assert(lowest([1, 2, 0]) === 3);
